linker.ts

```
@@ -1,72 +1,132 @@
+ * replaces the names with the provided hex address.
+ *
+ * @param bytecode
+ * @param hexAddress
+ * @param labelName
+ */
+function replacePlaceholderLabelWithHexAddress (bytecode, hexAddress, labelName) {
+  // truncate to 36 characters, why, I assume this is because legacy names
+  // only had a length of 36 characters?
+  const truncatedName = labelName.slice(0, 36);
+
+  const prefix = '__';
+  const postFix = `${Array(37 - truncatedName.length).join('_')}__`;
+  const libLabel = `${prefix}${truncatedName}${postFix}`;
+
+  while (bytecode.indexOf(libLabel) >= 0) {
+    bytecode = bytecode.replace(libLabel, hexAddress);
+  }
+
+  return bytecode;
+}
+
+/**
+ * When using libraries, the resulting bytecode will contain placeholders for
+ * the real addresses of the referenced libraries. These have to be updated,
+ * via a process called linking, before deploying the contract.
+ *
+ * This method provides a simple helper for linking.
+ *
+ * @param {string} bytecode
+ * @param {string} libraries
+ * @returns {string} bytecode
+ */
+function linkBytecode (bytecode, libraries) {
   assert(typeof bytecode === 'string');
   assert(typeof libraries === 'object');
+
   // NOTE: for backwards compatibility support old compiler which didn't use file names
   const librariesComplete = {};
-  for (const libraryName in libraries) {
-    if (typeof libraries[libraryName] === 'object') {
-      // API compatible with the standard JSON i/o
-      for (const lib in libraries[libraryName]) {
-        librariesComplete[lib] = libraries[libraryName][lib];
-        librariesComplete[libraryName + ':' + lib] = libraries[libraryName][lib];
-      }
-    } else {
-      // backwards compatible API for early solc-js versions
-      const parsed = libraryName.match(/^([^:]+):(.+)$/);
-      if (parsed) {
-        librariesComplete[parsed[2]] = libraries[libraryName];
+
+  for (const [libraryName, library] of Object.entries(libraries)) {
+    if (isNil(library)) continue;
```

I think that this used to fall into the `else` part and we'd have `librariesComplete[libraryName] = library` executed in that case. Now it's skipped.

Are we missing a test that would catch this?

---

Good point. But in that case I think we should just throw an error here instead of silently filtering it out.

---

What error would you like (text wise) here?